File: pulsar-skeleton/src/main/java/org/warps/pulsar/crawl/index/IndexWriter.java
package org.warps.pulsar.crawl.index;

import org.apache.hadoop.conf.Configuration;
import org.slf4j.Logger;
import org.slf4j.LoggerFactory;
import org.warps.pulsar.common.config.ReloadableParameterized;

import java.io.IOException;

/**
 * Created by vincent on 16-8-1.
 */
public interface IndexWriter extends ReloadableParameterized, AutoCloseable {

    Logger LOG = LoggerFactory.getLogger(IndexWriter.class);

    default String getName() {
        return getClass().getSimpleName();
    }

    default boolean isActive() {
        return true;
    }

    void open(Configuration conf) throws IOException;

    void open(String indexerUrl) throws IOException;

    void write(IndexDocument doc) throws IOException;

    void delete(String key) throws IOException;

    void update(IndexDocument doc) throws IOException;

    void commit() throws IOException;

    @Override
    void close() throws IOException;

    /**
     * Returns a String describing the IndexWriter instance and the specific
     * parameters it can take
     */
    String describe();
}

File: pulsar-skeleton/src/main/kotlin/ai/platon/pulsar/skeleton/crawl/index/IndexWriter.kt
package ai.platon.pulsar.skeleton.crawl.index

import ai.platon.pulsar.common.config.ImmutableConfig
import ai.platon.pulsar.common.config.Parameterized
import ai.platon.pulsar.skeleton.crawl.common.JobInitialized
import org.slf4j.LoggerFactory
import java.io.IOException

/**
 * Created by vincent on 16-8-1.
 */
interface IndexWriter : Parameterized, JobInitialized, AutoCloseable {
    val name: String get() = javaClass.simpleName
    val isActive: Boolean
        get() = true

    @Throws(IOException::class)
    fun open(conf: ImmutableConfig?)

    @Throws(IOException::class)
    fun open(indexerUrl: String?)

    @Throws(IOException::class)
    fun write(doc: IndexDocument?)

    @Throws(IOException::class)
    fun delete(key: String?)

    @Throws(IOException::class)
    fun update(doc: IndexDocument?)

    @Throws(IOException::class)
    fun commit()

    @Throws(IOException::class)
    override fun close()

    /**
     * Returns a String describing the IndexWriter instance and the specific
     * parameters it can take
     */
    fun describe(): String?

    companion object {
        val LOG = LoggerFactory.getLogger(IndexWriter::class.java)
    }
}